src/ModelContextProtocol/Client/AutoDetectingClientSessionTransport.cs

```diff
@@ -0,0 +1,200 @@
+/// </summary>
+internal sealed partial class AutoDetectingClientSessionTransport : ITransport
+{
+    private readonly SseClientTransportOptions _options;
+    private readonly HttpClient _httpClient;
+    private readonly ILoggerFactory? _loggerFactory;
+    private readonly ILogger _logger;
+    private readonly string _name;
+    private readonly DelegatingChannelReader<JsonRpcMessage> _delegatingChannelReader;
+    
+    private StreamableHttpClientSessionTransport? _streamableHttpTransport;
+    private SseClientSessionTransport? _sseTransport;
+
+    public AutoDetectingClientSessionTransport(SseClientTransportOptions transportOptions, HttpClient httpClient, ILoggerFactory? loggerFactory, string endpointName)
+    {
+        Throw.IfNull(transportOptions);
+        Throw.IfNull(httpClient);
+
+        _options = transportOptions;
+        _httpClient = httpClient;
+        _loggerFactory = loggerFactory;
+        _logger = (ILogger?)loggerFactory?.CreateLogger<AutoDetectingClientSessionTransport>() ?? NullLogger.Instance;
+        _name = endpointName;
+        _delegatingChannelReader = new DelegatingChannelReader<JsonRpcMessage>(this);
+    }
+
+    /// <summary>
+    /// Returns the active transport (either StreamableHttp or SSE)
+    /// </summary>
+    internal ITransport? ActiveTransport => _streamableHttpTransport != null ? (ITransport)_streamableHttpTransport : _sseTransport;
+
+    public ChannelReader<JsonRpcMessage> MessageReader => _delegatingChannelReader;
+
+    /// <inheritdoc/>
+    public async Task SendMessageAsync(JsonRpcMessage message, CancellationToken cancellationToken = default)
+    {
+        if (_streamableHttpTransport == null && _sseTransport == null)
+        {
+            var rpcRequest = message as JsonRpcRequest;
+            
+            // The first message must be an initialize request
+            Debug.Assert(rpcRequest != null && rpcRequest.Method == RequestMethods.Initialize, 
+                "First message must be an initialize request");
+
+            // Try StreamableHttp first
+            _streamableHttpTransport = new StreamableHttpClientSessionTransport(_options, _httpClient, _loggerFactory, _name);
+            
+            try
+            {
+                LogAttemptingStreamableHttp(_name);
+                var response = await _streamableHttpTransport.SendInitialRequestAsync(message, cancellationToken).ConfigureAwait(false);
+                
+                // If the status code is not success, fall back to SSE
+                if (!response.IsSuccessStatusCode)
+                {
+                    LogStreamableHttpFailed(_name, response.StatusCode);
+                    
+                    try
+                    {
+                        await _streamableHttpTransport.DisposeAsync().ConfigureAwait(false);
+                    }
+                    finally
+                    {
+                        _streamableHttpTransport = null;
```

Comment:
Don't bother nulling out _streamableHttpTransport except when nothing throws. We only need to handle the common, non-200 status code failure. Any other failure we should just let the exception bubble up and dispose any transports, but we don't need to assign them to null except in the normal IsSuccessStatusCode case when ._streamableHttpTransport.DisposeAsync() doesn't throw.